truffle/src/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/ReachabilityTest.java

```diff
@@ -301,4 +304,62 @@ int do1() throws RuntimeException {
 
     }
 
+    abstract static class ReachabilityUncached extends Node {
+        abstract int execute();
+
+        int foo() {
+            return 0;
+        }
+    }
+
+    @GenerateUncached
+    abstract static class ReachabilityUncached1 extends ReachabilityUncached {
+        @Specialization
+        int doCached(@Cached("foo()") int cached) {
+            return cached;
+        }
+
+        @Specialization(replaces = "doCached")
+        int doUncached() {
+            return 1;
+        }
+    }
+
+    @GenerateUncached
+    abstract static class ReachabilityUncached2 extends ReachabilityUncached {
```

Comment:
Here is a failing test.

```

 @GenerateUncached
    abstract static class ReachabilityUncached4 extends ReachabilityUncached {

        @Specialization
        int s0(@Cached("foo()") int cached) {
            return cached;
        }

        @Specialization(replaces = "s0")
        int s1() {
            return 1;
        }

        @Specialization(replaces = "s1")
        int s2() {
            return 1;
        }
    }

```

I think that one should work too but doesn't.

Reply:
I don't understand what's wrong with this example. It fails with `Specialization is not reachable. It is shadowed by s0(int).`. And that is true – `s1` is replaced and as such is never a part of the uncached node. So normal rules apply here (as if `@GenerateUncached` was not present).

Reply:
You are right. I tried somehow to construct a case where the missing handling of transitivity was a problem. But it seems it never was.